### ISSUE TITLE
secondary is not recognized as being down

### ISSUE THREAD

**Issue by s4ke** [citusdata/pg_auto_failover#232]
I have the following setup on the monitor:

```
postgres@af-monitor-01 /var/lib/postgresql/12/main_cluster $ pg_autoctl show state --pgdata /var/lib/postgresql/12/main_cluster/
          Name |   Port | Group |  Node |     Current State |    Assigned State
---------------+--------+-------+-------+-------------------+------------------
<ip> |   5433 |     0 |     1 |         secondary |         secondary
    <ip> |   5433 |     0 |     2 |           primary |           primary
   <ip> |   5433 |     1 |     3 |            single |            single
```

pg_hba.conf on all servers:

```
local all postgres    peer
local template1 autoctl_node    peer map=autoctl_node_peer
local pg_auto_failover autoctl_node    peer map=autoctl_node_peer
hostssl autoctl autoctl_node <ip>/32   cert map=autoctl_node_remote
hostssl autoctl autoctl_node <ip>/32   cert map=autoctl_node_remote
hostssl autoctl autoctl_node <ip>/32   cert map=autoctl_node_remote
hostssl autoctl autoctl_node <ip>/32   cert map=autoctl_node_remote
hostssl template1 autoctl_node <ip>/32   cert map=autoctl_node_remote
hostssl template1 autoctl_node <ip>/32   cert map=autoctl_node_remote
hostssl template1 autoctl_node <ip>/32   cert map=autoctl_node_remote
hostssl template1 autoctl_node <ip>/32   cert map=autoctl_node_remote
hostssl pg_auto_failover autoctl_node <ip>/32   cert map=autoctl_node_remote
hostssl pg_auto_failover autoctl_node <ip>/32   cert map=autoctl_node_remote
hostssl pg_auto_failover autoctl_node <ip>/32   cert map=autoctl_node_remote
hostssl pg_auto_failover autoctl_node <ip>/32   cert map=autoctl_node_remote
hostssl replication pgautofailover_replicator <ip>/32   cert map=pgautofailover_replicator_remote
hostssl replication pgautofailover_replicator <ip>/32   cert map=pgautofailover_replicator_remote
hostssl replication pgautofailover_replicator <ip>/32   cert map=pgautofailover_replicator_remote
hostssl replication pgautofailover_replicator <ip>/32   cert map=pgautofailover_replicator_remote
```

pg_ident.conf on all servers:

```
autoctl_node_peer  postgres             autoctl_node
autoctl_node_remote <ip> autoctl_node
pgautofailover_replicator_remote <ip> pgautofailover_replicator
autoctl_node_remote <ip> autoctl_node
pgautofailover_replicator_remote <ip> pgautofailover_replicator
autoctl_node_remote <ip> autoctl_node
pgautofailover_replicator_remote <ip> pgautofailover_replicator
autoctl_node_remote <ip> autoctl_node
pgautofailover_replicator_remote <ip> pgautofailover_replicator
```

We are initializing the monitor with this command (templated with Ansible):

```
    PATH="$PATH:/usr/lib/postgresql/12/bin" pg_autoctl create monitor \
    --pgdata "{{ postgresql_data_dir }}" \
    --skip-pg-hba \
    --ssl-ca-file "{{ postgresql_cluster_ssl_ca_file | default('/data/ansible/certs/postgres_server/rootCA.crt') }}" \
    --server-key "{{ postgresql_cluster_server_key | default('/data/ansible/certs/postgres_server/server.key') }}" \
    --server-cert "{{ postgresql_cluster_server_cert | default('/data/ansible/certs/postgres_server/server.crt') }}" \
    --nodename "{{ host_ip }}" \
    --pgport "{{ postgresql_cluster_port | default('5433') }}"
```

We are initializing our database nodes with this command:

```
    PATH="$PATH:/usr/lib/postgresql/12/bin" pg_autoctl create postgres \
    --pgdata "{{ postgresql_data_dir }}" \
    --skip-pg-hba \
    --ssl-ca-file "{{ postgresql_cluster_ssl_ca_file | default('/data/ansible/certs/postgres_server/rootCA.crt') }}" \
    --server-key "{{ postgresql_cluster_server_key | default('/data/ansible/certs/postgres_server/server.key') }}" \
    --server-cert "{{ postgresql_cluster_server_cert | default('/data/ansible/certs/postgres_server/server.crt') }}" \
    --nodename "{{ host_ip }}" \
    --pgport "{{ postgresql_cluster_port | default('5433') }}" \
    --monitor postgres://autoctl_node@{{ first_monitor_node }}:{{ postgresql_cluster_port | default('5433') }}/pg_auto_failover
```

The initial cluster looks fine to me according to the state shown by the monitor is fine (ignore the third single node, this was just me trying out what happens if i add a third node).

Now, in order to test the resilience of this setup I tried the following:

- demoted node 1 to secondary (worked just like the documentation told me)
- essentially *deleted* node 1 by recreating the complete Cloud Virtual machine

What I expected:

- The monitor should register that the secondary is down, move the primary to state WAIT_PRIMARY

What actually happened:

- nothing
- running statements (`CREATE database test`) locked up psql, when force closing I get the message:

```
postgres=# create database test;
^CCancel request sent
WARNING:  canceling wait for synchronous replication due to user request
DETAIL:  The transaction has already committed locally, but might not have been replicated to the standby.
```

Versions:

- PostgreSQL 12.2 (Ubuntu 18.04)

pg_autoctl:

```
pg_autoctl version 1.2
compiled with PostgreSQL 12.1 (Ubuntu 12.1-1.pgdg18.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0, 64-bit
compatible with Postgres 10, 11, and 12
```

**Comment by s4ke** [comment]
Also, this is particularly problematic because I managed to run a switchover to that node and the monitor happily accepted my request.

**Comment by s4ke** [comment]
Sidenote: This might have happened because of previous unsuccessful attempts of creating the cluster.

**Comment by s4ke** [comment]
If it helps, we are using the following scripts:

https://github.com/neuroforgede/pg_auto_failover_ansible

**Comment by DimCitus** [comment]
Hi @s4ke ; thanks for your interest in pg_auto_failover!

While the setup of your environment is very clearly documented in your issue (thanks for that!), the operations that you implemented then in the cluster are not detailed at all. I don't know what you tried and what you observed in details. Can you provide specific details about what you did and the result you obtained?

In particular we do not have an operation that allows a user to "demote a node to secondary". What do you do? We don't have a way to "essentially delete a node". What did you do?

**Comment by s4ke** [comment]
Hey @DimCitus ,

ok, let me try again :)

Starting off with a freshly installed cluster I did the following (this is as I remember it):

1. Issued a switchover via pg_autoctl by running on the monitor node:

```
pg_autoctl --pgdata <path-to-pg-data> perform switchover
```

This worked. 

2. But then, I went to my cloud provider (Hetzner) and tried to simulate a worst case scenario. I rebuilt the VM (full disk wipe, reinstallation of Ubuntu 18.04 from scratch on the same IP) of the now secondary node - this was the primary before step 1 - without removing it from the cluster.

What I expected:

- I expected the monitor node to recognize that the secondary node was not reachable anymore and switch into the state WAIT_PRIMARY. 

What happened instead:

- But instead no state changes occured.
- Additionally, all writes on the primary node locked up the cli as described in the initial posting(s).

Further things I tried:

- To see what happened if I tried a switchover, ran the pg_autoctl command again and the monitor then tried to switch the now broken secondary to the new primary node essentially frying the cluster.

If you want, I can try to reproduce these issues again, even in a live session if you want.

**Comment by s4ke** [comment]
I will check this again against version 1.3 and see if the errors persists.

**Comment by DimCitus** [comment]
Hi @s4ke ; thanks for providing further input. I'm curious as to why the monitor would fail to realise that your secondary was missing. To decide that the monitor uses health checks, every once in a while it connects to the Postgres instance on the secondary. Those should have been failing at some point in your VM rebuild process I'm sure. Can you share Postgres logs from the monitor database, where we should see the health checks failing?

**Comment by s4ke** [comment]
Sorry, currently I do not have the time to follow this up as promised.

I don't have the postgres logs from the monitor anymore, but I found some slight irregularities regarding pg_hba settings, that might have been the root cause of this issue.

**Comment by DimCitus** [comment]
Thanks for feedback. I'm now closing this issue then, for triaging. If you have more information or can reproduce the issue, feel free to re-open then, of course.
